---
title: "R Notebook"
output: html_notebook
---

```{r setup, echo = F}
knitr::opts_chunk$set(echo = F, message = F, warning = F)
```

# load required packages

```{r load-packages}
library(glue)
library(fs)
library(rprojroot)
library(tidyverse)
proj_dir = rprojroot::find_root(criterion = has_file_pattern("*.Rproj"))
cohorts = c("vc", "reynolds") %>% 
  set_names(.)

library(reshape2)
library(data.table)
library(karyoploteR)
library(scales)
library(biobroom)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene #shorthand (for convenience)
library(org.Hs.eg.db)
library(plyranges)
library(stchlkExome)
library(patchwork)
```

# load segmentation data

```{r, eval = F}
seg_granges <- vector(mode = "list", length = 2)
```

```{r, eval = F}
# load known genes
hg19_genes <- genes(txdb)

# load segmentation data ---------------------------------------------------------------

# load kooi SCNA peak regions 
kooi_peak_regions <- read_csv("~/rb_pipeline/doc/SCNA/kooi_SCNA_peak_regions.csv")

kooi_peak_granges <- makeGRangesFromDataFrame(kooi_peak_regions)

segmentation_files <- list(
  "vc" = c("~/rb_pipeline/output/copywriter/vc/50kb/CNAprofiles/",
           "~/rb_pipeline/output/copywriter/vc/50kb_additional/CNAprofiles/"),
  "reynolds" = c("~/rb_pipeline/output/copywriter/reynolds/50kb/CNAprofiles/")
)


```

# collate segementation data

```{r, eval = FALSE}

seg_granges <- purrr::map(cohorts, collate_scna_segments, segmentation_files)

seg_granges_33_N <- seg_granges$vc$`33.N`

seg_granges$vc <- seg_granges$vc[!grepl("N", names(seg_granges$vc))]
seg_granges$vc <- sapply(seg_granges$vc, function(x) x[!grepl("none", x$ID)])

seg_granges$vc <- c(seg_granges$vc[1:14], `33.N` = seg_granges_33_N, seg_granges$vc[15:24])

karyo_segs <- purrr::map(cohorts, collate_scna_segments, segmentation_files, as_grange = FALSE)

saveRDS(seg_granges, "~/rb_pipeline/results/SCNA/seg_granges.rds")
saveRDS(karyo_segs, "~/rb_pipeline/results/SCNA/karyo_segs.rds")
```

# calculate baf granges

```{r, eval = FALSE }
baf_granges <- vector(mode = "list")

# load kooi SCNA peak regions 
kooi_peak_regions <- read_csv("~/rb_pipeline/doc/SCNA/kooi_SCNA_peak_regions.csv")

kooi_peak_granges <- makeGRangesFromDataFrame(kooi_peak_regions)

mbaf_threshold = 0.65
```

```{r, eval = F}
cohort = "vc"

# load input files when processed per sample -------------------------------
# baf_segment_files <- list.files("~/rb_pipeline/output/bafsegmentation/segmented", recursive = TRUE, pattern = "all_AI_regions.txt", full.names = TRUE)
# 
# baf_segment_files <- c(
#   "/home/skevin/rb_pipeline/output/bafsegmentation/segmented/33-CL/AI_regions.txt",
#   "/home/skevin/rb_pipeline/output/bafsegmentation/segmented/43-T/AI_regions.txt",
#   "/home/skevin/rb_pipeline/output/bafsegmentation/segmented/all_AI_regions.txt"
# )  

baf_segment_files <-
  fs::path("~/rb_pipeline/output/bafsegmentation/segmented/") %>%
  dir_ls(recurse = TRUE) %>%
  path_filter(regex = ".*/[0-9]{2}-(T|CL)/AI_regions.txt") %>%
  identity()

# baf_segment_files <- head(baf_segment_files, -2)

names(baf_segment_files) <- path_file(path_dir(baf_segment_files))

# # for combining separate rounds of bafsegmentation
# addl_samples <- strsplit(baf_segment_names[1:2], "-")
# addl_samples <- sapply(addl_samples, function(x) paste0(x[2], "_", x[1]))
#   
baf_segment_list <- lapply(baf_segment_files, read.table, header = TRUE)
baf_segment_list <- lapply(baf_segment_list, clean_baf_segment)
# 
# multi_baf <- split(baf_segment_list[[3]], baf_segment_list[[3]]$Assay)
# baf_segment_list <- c(baf_segment_list[1], baf_segment_list[2], multi_baf)
# baf_segment_names[1:2] <- addl_samples

baf_granges$vc <- map(baf_segment_list, makeGRangesFromDataFrame, keep.extra.columns = TRUE)
# names(baf_granges$vc) <- sapply(baf_granges$vc, function(x) unique(x$Assay))

new_order <- tibble::tibble(
  sampleid = names(baf_granges$vc), 
  samplenumber = stringr::str_extract(names(baf_granges$vc), "^[0-9]*")) %>%
  dplyr::group_by(samplenumber) %>%
  dplyr::arrange(desc(sampleid), .by_group = TRUE) %>%
  dplyr::pull(sampleid) %>%
  identity()

baf_granges$vc <- baf_granges$vc[new_order]
# saveRDS(baf_granges$vc, "~/rb_pipeline/doc/LOH/LOH_vc_granges.rds")

# Allelic Imbalance called LOH when mBAF > ?; threshold is set at 0.7 in Kooi paper
baf_granges$vc <- purrr::map(baf_granges$vc, ~plyranges::filter(.x, mBAF > mbaf_threshold))

```

```{r, eval = F}
cohort = "reynolds"

baf_segment_files <- list.files("~/rb_pipeline/output/bafsegmentation/segmented/", recursive = TRUE, pattern = "AI_regions.txt", full.names = TRUE)

# filter for reynolds baf segment files
baf_segment_files <- baf_segment_files[grepl("\\d{3}-CL/", baf_segment_files)]

baf_segment_names <- basename(dirname(baf_segment_files))

baf_segment_list <- map(baf_segment_files, read.table, header = TRUE)
baf_segment_list <- map(baf_segment_list, clean_baf_segment)

baf_granges$reynolds <- map(baf_segment_list, makeGRangesFromDataFrame, keep.extra.columns = TRUE)

names(baf_granges$reynolds) <- sapply(baf_granges$reynolds, function(x) unique(x$Assay))

saveRDS(baf_granges$reynolds, "~/rb_pipeline/doc/LOH/LOH_reynolds_granges.rds")

# Allelic Imbalance called LOH when mBAF > ?; threshold is set at 0.7 in Kooi paper
baf_granges$reynolds <- purrr::map(baf_granges$reynolds, ~plyranges::filter(.x, mBAF > mbaf_threshold))


baf_granges <- purrr::map(baf_granges, ~set_names(.x, str_replace(names(.x), "-", ".")))

saveRDS(baf_granges, "~/rb_pipeline/results/LOH/baf_granges.rds")

```

## load segmentation from rds

```{r }



baf_granges <- readRDS("~/rb_pipeline/results/LOH/baf_granges.rds")

seg_granges <- readRDS("~/rb_pipeline/results/SCNA/seg_granges.rds")

karyo_segs <- readRDS("~/rb_pipeline/results/SCNA/karyo_segs.rds") 

karyo_seg_relative <- karyo_segs$vc %>% 
  dplyr::filter(!str_detect(ID, "none")) %>% 
  dplyr::filter(!str_detect(ID, ".*N.*N.*"))

karyo_seg_33_N <- dplyr::filter(karyo_segs$vc, ID == "log2.33.N_1_recalibrated.bam.vs.none")

karyo_segs$vc <- dplyr::bind_rows(karyo_seg_relative)

purrr::imap(karyo_segs, ~write_tsv(.x, paste0("~/rb_pipeline/results/SCNA/", .y, ".seg")))


```

# plot baf segmentation data all chromosomes

```{r, message = FALSE}
# plot vc seg

vc_ai_plot <-
  ggplotify::as.ggplot(expression(plot_loh_granges(baf_granges$vc, chr_select = "auto", marker_granges = kooi_peak_granges, data2height = 100, bottommargin = 0.1, topmargin = 10, lwd = 0.01, leftmargin = 0.05)))


# output vc seg             
vc_baf_path <- "~/rb_pipeline/results/LOH/vc_loh_segments.txt"
# output table of Allelic Imbalance Regions
vc_baf_regions <- dplyr::bind_rows(lapply(baf_granges$vc, data.frame))
write.csv(vc_baf_regions, vc_baf_path)

# plot reynolds seg
reynolds_ai_plot <-
  ggplotify::as.ggplot(expression(plot_loh_granges(baf_granges$reynolds, chr_select = "auto", marker_granges = kooi_peak_granges, data2height = 100, bottommargin = 0.1, topmargin = 12, lwd = 0.01, leftmargin = 0.05)))

# output table of Allelic Imbalance Regions
reynolds_baf_path <- "~/rb_pipeline/results/LOH/reynolds_loh_segments.txt"
reynolds_baf_regions <- dplyr::bind_rows(lapply(baf_granges$reynolds, data.frame))
write.csv(reynolds_baf_regions, reynolds_baf_path)

```

# plot scna segmentation data all chromosomes

```{r, message = FALSE}
# plot vc seg

vc_scna_plot <-
  ggplotify::as.ggplot(expression(plot_scna_granges(seg_granges$vc, chr_select = "auto", marker_granges = kooi_peak_granges, data2height = 100, bottommargin = 0.1, topmargin = 10, lwd = 0.01, leftmargin = 0.05)))

# plot reynolds seg
reynolds_scna_plot <-
  ggplotify::as.ggplot(expression(plot_scna_granges(seg_granges$reynolds, 
                                                    chr_select = "auto", 
                                                    marker_granges = kooi_peak_granges, 
                                                    data2height = 100, 
                                                    bottommargin = 0.1, 
                                                    topmargin = 12, 
                                                    lwd = 0.01, 
                                                    leftmargin = 0.05))) + 
  theme(plot.margin = margin(0, 0, 0, 0, "cm"))


```

```{r }

merged_heatmap_file <- "~/rb_pipeline/doc/RB_exome_manuscript/stachelek_supplemental/fig_03.pdf"

final_plot <- vc_scna_plot / 
  reynolds_scna_plot / 
  vc_ai_plot / 
  reynolds_ai_plot / 
  # plot_layout(heights = c(1.5, 1, 1.5, 1)) +
  plot_annotation(tag_levels = 'A') &
  theme(plot.tag = element_text(size = 8))

ggsave(merged_heatmap_file, final_plot, height = 6, width = 6)

```

# scale bars for plots

```{r eval = F}

SCNA_scale_plot <- function(scna_df) {
  # browser()
  
  max_scale = max(scna_df$seg.mean)
  
  scna_df <- 
    scna_df %>% 
    dplyr::mutate(seg.mean = pmax(seg.mean, -2)) %>%
    # dplyr::mutate(seg.mean = scales::rescale_mid(seg.mean, to = c(-2, max_scale), mid = 0)) %>%
    # dplyr::mutate(seg.mean = 2*2^(seg.mean)) %>%
    identity()
  
  p1 <- ggplot(scna_df, aes(x = sample_id, y = seg.mean)) + geom_point(aes(color = seg.mean)) + 
    # scale_color_gradient2(guide = guide_colourbar(), low = "blue", mid = "white", high = "red", midpoint = 0) +
    scale_color_gradientn(
      colors=c("blue","white","red"),
      values=rescale(c(-2,0, max_scale)),
      limits=c(-2,max_scale)
    ) +
    # guides(fill = guide_colourbar()) + 
    NULL
}

scale_bar_plots <- purrr::map(karyo_segs, SCNA_scale_plot)

scale_plot_paths <- paste0("~/rb_pipeline/results/SCNA/", cohorts, "_scales.pdf")

purrr::map2(scale_plot_paths, scale_bar_plots, ggsave)

```


# targeted SCNA plots

## 8a

For example, tumor CHLA-VC-RB20 had subclonal 16q loss while the cell line had clonal 16q loss, de novo multi-copy 2p gain, and gain of additional copies of 1q (Fig. 5A); 

tumor CHLA-VC-RB41 had subclonal 6p gain and 16q loss whilst the cell line had clonal 6p gain and 16q loss as well as subclonal 2p gain and focal MYCN amplification (Figure 9); 

tumor CHLA-VC-RB49 had subclonal 1q+, 6p+, 16q- whilst the cell line had clonal 1q+, 6p+, 16q-, as well as 2p+.

MYCN-amplified CHLA-VC-RB24 tumor yielded a cell line with loss of the RB1-containing 13q (Fig. 5B) and pRB expression, implying emergence of a RB1+/-, MYCNA subclone. 

As an exception, a nearly clonal 16q loss in CHLA-VC-RB46 was restored in the cell line, indicative of selection of a rare subclone that retained both 16q copies and likely had genomic changes that were highly advantageous in culture. 

Another tumor, CHLA-VC-RB28, reverted a 2p gain yet had 2p CN-LOH indicative of two copies of the initially gained 2p allele. Summary sentence??

```{r }

callout_samples <- c("14.T", "14.CL", "20.T", "20.CL", "24.T", "24.CL", "28.T", "28.CL", "31.T", "31.CL", "41.T", "41.CL", "46.T", "46.CL", "49.T", "49.CL")

fig_8 <- ggplotify::as.ggplot(expression(plot_scna_granges(seg_granges$vc[callout_samples], chr_select = c("chr1", "chr2", "chr6", "chr13", "chr16", "chrX"), data2height = 1, ideogramheight = 0.05, topmargin = 0.1, data1height = 0.05, data1inmargin = 0, data1outmargin = 0, data2inmargin = 0.05, cex = 0.75, lwd = 0.01)))

ggsave("~/rb_pipeline/doc/RB_exome_manuscript/stachelek_supplemental/fig_04.pdf", fig_8, height = 6, width = 18)


```


### RB46 T vs CL completely different?

# find overlapping genes in gain/loss regions samples

```{r}

cohort = "vc"

hg19_genes <- genes(txdb)

seg_granges$vc <- c(seg_granges$vc, `33.N` = seg_granges_33_N)

seg_granges[[cohort]] <- purrr::map(seg_granges[[cohort]], ~find_overlaps(hg19_genes, .x))

# test filtering with -0.5 loss
all_scnas <- seg_granges[[cohort]] %>% 
  purrr::map(~as_tibble(.x)) %>% 
  dplyr::bind_rows() %>%
  dplyr::mutate(gene_id = as.integer(gene_id)) %>%
  dplyr::group_by(gene_id, sample_id, chromosome = seqnames) %>% 
  dplyr::summarize(seg.mean = mean(seg.mean)) %>%
  dplyr::mutate(seg.mean = 2*2^(seg.mean)) %>%
  identity()
  

gains <- dplyr::filter(all_scnas, seg.mean > 2.5)

losses <- dplyr::filter(all_scnas, seg.mean < 1.5)


find_gain_loss_genes <- function(scna_tbl, hg19_genes){
  
  scna_tbl <- scna_tbl %>% 
  tidyr::spread(sample_id, seg.mean) %>%
  identity()

  scna_tbl <- annotables::grch37 %>% 
    select(entrez, symbol) %>% 
    dplyr::inner_join(scna_tbl, by = c("entrez" = "gene_id")) %>%
    dplyr::select(-entrez) %>% 
    identity()
}

all_scnas <- find_gain_loss_genes(all_scnas, hg19_genes)

gains <- find_gain_loss_genes(gains, hg19_genes)

losses <- find_gain_loss_genes(losses, hg19_genes)

write.csv(all_scnas, glue::glue("~/rb_pipeline/doc/RB_exome_manuscript/stachelek_supplemental/table_s18.csv"))


write.csv(losses, glue::glue("~/rb_pipeline/doc/RB_exome_manuscript/SCNA/{cohort}_losses.csv"))
write.csv(gains, glue::glue("~/rb_pipeline/doc/RB_exome_manuscript/SCNA/{cohort}_gains.csv"))


```

# compare gains/losses with results from Kooi meta-analysis

```{r}
# Besides the well-established driver genes RB1 (13q-loss) and MYCN (2p-gain) we identified CRB1 and NEK7 (1q-gain), SOX4 (6p-gain) and NUP205 (7q-gain)  alternative candidates were identified including MIR181 (1q-gain) and DEK (6p gain)

kooi_genes <- c("CRB1", "NEK7", "SOX4", "NUP205")

kooi_overlap_scnas <- purrr::map(seg_granges[[cohort]], ~find_overlaps(kooi_peak_granges, .x)) 

kooi_overlap_scnas <- kooi_overlap_scnas %>% 
  # map(~plyranges::filter(.x, seg.mean < -0.5)) %>%
  purrr::map(~as_tibble(.x)) %>% 
  dplyr::bind_rows() %>%
  dplyr::mutate(gene_id = as.integer(gene_id)) %>%
  dplyr::group_by(gene_id, sample_id, chromosome = seqnames) %>% 
  dplyr::summarize(seg.mean = mean(seg.mean)) %>%
  dplyr::mutate(seg.mean = 2*2^(seg.mean)) %>%
  identity()


kooi_overlap_scnas <- find_gain_loss_genes(kooi_overlap_scnas, hg19_genes)

kooi_overlap_gains <- dplyr::filter(kooi_overlap_scnas, chromosome %in% c("chr1", "chr2", "chr6", "chr7", "chr19"))

kooi_overlap_losses <- dplyr::filter(kooi_overlap_scnas, chromosome %in% c("chr13", "chr16"))



```

```{r }
mtcars2 <- within(mtcars, {
  vs <- factor(vs, labels = c("V-shaped", "Straight"))
  am <- factor(am, labels = c("Automatic", "Manual"))
  cyl  <- factor(cyl)
  gear <- factor(gear)
})

p1 <- ggplot(mtcars2) +
  geom_point(aes(x = wt, y = mpg, colour = gear)) +
  labs(title = "Fuel economy declines as weight increases",
       subtitle = "(1973-74)",
       caption = "Data from the 1974 Motor Trend US magazine.",
       tag = "Figure 1",
       x = "Weight (1000 lbs)",
       y = "Fuel economy (mpg)",
       colour = "Gears")

p1 + theme_gray() # the default
```


```{r }
p1 + theme_bw()
```


```{r }
p1 + theme_linedraw()
```


```{r }
p1 + theme_light()
```


```{r }
p1 + theme_dark()
```


```{r }
p1 + theme_minimal()
```


```{r }
p1 + theme_classic()
```


```{r }
p1 + theme_void()
```