---
title: "Report with text"
author: "Kevin Stachelek"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message=FALSE)
```

# load packages

```{r}
source("../../functions.R")
source("../../packages.R")

library(tidyverse)
library(drake)
library(glue)
library(english)
```

## Analysis

```{r}

loadd(vc_vars)

loadd(vaf_plot_input)

loadd(filtered_vaf_plot_input)

loadd(reynolds_snv)

loadd(all_study_coverage)

loadd(noncoding_all_study_snvs)

loadd(all_study_mutations)

loadd(afshar_mutations)

loadd(prior_study_mutations)

loadd(prior_study_snvs)

loadd(all_study_snv_qc)
```


# load Seiler et al. 119 canonical splicing factor genes

```{r}

loadd(all_study_mutations_formatted)

seiler_119 <- read_csv("~/rb_pipeline/doc/dflow_output/seiler_119_genes.csv") %>% 
  janitor::clean_names() %>% 
  dplyr::filter(!is.na(final_119_and_reason_for_inclusion)) %>%
  dplyr::pull(gene_symbol) %>% 
  identity()

intersect(seiler_119, all_study_mutations_formatted$gene)




```

### "Apart from BCOR, CREBBP, and MYCN, we and others have not detected variants in genes that were found to be recurrently altered in targeted sequencing studies"

```{r}

# "BCOR", "MYCN", "CREBBP"

francis_genes <- c("FAT1", "RPTOR", "ARID1A", "MSH3", "TERT", "TSC2")

targeted_reported_genes <- 
  c(afshar_mutations$gene, francis_genes) %>% 
  unique()

test0 <- 
all_study_mutations %>% 
  dplyr::filter(gene %in% targeted_reported_genes) %>% 
  dplyr::filter(gene != "BCOR")

```


```{r}
n_samples <- 
all_study_snv_qc$vars_per_study %>% 
  dplyr::distinct(sample_set, sample) %>% 
  dplyr::group_by(sample_set) %>% 
  dplyr::count() %>% 
  identity()

n_ngs_samples <- 
n_samples %>% 
  dplyr::filter(!str_detect(sample_set, "Stachelek|Afshar")) %>% 
  identity()

ngs_mutations <-
  all_study_mutations %>%
  dplyr::filter(sequencing_format %in% c("WES", "WGS"))

n_BCOR_all <- 
  ngs_mutations %>% 
  dplyr::filter(gene == "BCOR") %>% 
  dplyr::filter(!str_detect(sample, "-CL")) %>% 
  dplyr::distinct(sample, .keep_all = TRUE) %>% 
  nrow() %>%
  identity()

n_CREBBP_all <- 
  ngs_mutations %>% 
  dplyr::filter(gene == "CREBBP") %>% 
  dplyr::distinct(sample) %>%
  nrow() %>% 
  identity()

prior_ngs_mutations <- 
  ngs_mutations %>% 
  dplyr::filter(!str_detect(study, "Stachelek"))

n_BCOR_prior <- 
  prior_ngs_mutations %>% 
  dplyr::filter(gene == "BCOR") %>% 
  dplyr::filter(!str_detect(sample, "-CL")) %>% 
  dplyr::distinct(sample, .keep_all = TRUE) %>% 
  nrow() %>%
  identity()

n_CREBBP_prior <- 
  prior_ngs_mutations %>% 
  dplyr::filter(gene == "CREBBP") %>% 
  dplyr::distinct(sample) %>%
  nrow() %>% 
  identity()
```


```{r}
prior_ngs_snvs <- 
  prior_study_snvs %>% 
  dplyr::filter(!study == "Afshar et al.")

recurrent_prior_ngs_snvs <- 
  prior_ngs_snvs %>% 
    dplyr::group_by(gene) %>% 
  dplyr::filter(dplyr::n() > 1)

private_prior_ngs_snvs <-
  prior_ngs_snvs %>% 
  dplyr::group_by(gene) %>% 
  dplyr::filter(dplyr::n() == 1)

all_ngs_snvs <- 
  noncoding_all_study_snvs %>% 
  dplyr::filter(!study == "Afshar et al.")
```


```{r}
prior_targeted_snvs <- 
  prior_study_snvs %>% 
  dplyr::filter(study == "Afshar et al.") %>% 
  dplyr::group_by(gene) %>%
  identity()

private_prior_targeted_snvs <- 
  prior_targeted_snvs %>% 
  group_by(gene) %>%
  dplyr::filter(dplyr::n() == 1)

n_private_prior_targeted_snvs <- 
  private_prior_targeted_snvs %>% 
  nrow() 

recurrent_prior_targeted_snvs <- 
  prior_targeted_snvs %>% 
  group_by(gene) %>% 
  dplyr::filter(dplyr::n() > 1) %>% 
  janitor::tabyl(gene) %>% 
  dplyr::pull(gene)

```


```{r}
recurrent_mutations <-
  all_study_mutations %>%
  dplyr::mutate(sample_number = stringr::str_remove(sample, "-.*")) %>% 
  dplyr::distinct(sample, gene, .keep_all = TRUE) %>%
  group_by(gene) %>%
  dplyr::filter(n_distinct(sample_number) > 1) %>%
  identity()

recurrent_ngs_snvs <- 
  recurrent_mutations %>% 
  dplyr::filter(!study == "Afshar et al.") %>% 
  dplyr::filter(!modality == "focal_scna")

recurrent_ngs_genes <-
  recurrent_ngs_snvs %>% 
  dplyr::pull(gene)

all_recurrent <- unique(c(recurrent_ngs_genes, recurrent_prior_targeted_snvs))
```


```{r}
stachelek_filtered_variants <- 
  all_study_snv_qc$vaf_per_study %>% 
  dplyr::filter(study == "Stachelek") %>% 
    dplyr::mutate(sample_type = case_when(grepl("T", sample) ~ "Tumor",
                                       grepl("CL", sample) ~ "Cell Line",
                                       grepl("N", sample) ~ "Normal")) %>% 
  identity()
```


```{r}
stachelek_filtered_cell_line_variants <- 
  stachelek_filtered_variants %>% 
  dplyr::filter(str_detect(sample, "-CL")) %>%
  dplyr::distinct(chr, start, end, ref, alt, .keep_all = TRUE) %>%
  # dplyr::arrange(sample, chr, start, end, ref, alt) %>% 
  dplyr::group_by(sample) %>%
  dplyr::count() %>%
  tibble::deframe() %>%
  identity()

stachelek_filtered_cell_line_only_variants <- 
  stachelek_filtered_variants %>% 
  dplyr::group_by(chr, start, end) %>% 
  dplyr::filter(!any(str_detect(sample, "T"))) %>% 
  dplyr::distinct(sample, chr, start, end, ref, alt) %>% 
  dplyr::group_by(sample) %>%
  dplyr::count() %>% 
  tibble::deframe() %>% 
  identity()

stachelek_filtered_tumor_variants <- 
  stachelek_filtered_variants %>% 
  dplyr::filter(str_detect(sample, "-T")) %>%
  dplyr::group_by(sample) %>%
  dplyr::count() %>% 
  ungroup() %>% 
  tibble::add_row(sample = "41-T", n = 0) %>% 
  tibble::deframe()

filtered_circle_ids <-
  find_selected_variants(filtered_vaf_plot_input) %>% 
  purrr::map(dplyr::filter, circle_id == "circled")
  
```

```{r}
# need to include intu and hrg recurences in kooi +2


glue("Additionally, whole genome or exome sequencing of {sum(n_ngs_samples$n)} treatment-naïve retinoblastomas revealed pathogenic mutation or focal deletion of BCOR in {n_BCOR_prior}, CREBBP variants in {english(n_CREBBP_prior)}, and somatic variants of {nrow(private_prior_ngs_snvs)+2} genes thus far reported for only one sample")

```

```{r}

loadd(prior_study_snvs)

kooi_mutations <- 
  prior_study_snvs %>% 
  dplyr::filter(study == "Kooi et al.")

non_kooi_mutations <- 
  prior_study_snvs %>% 
  dplyr::filter(!study %in% c("Kooi et al.", "Afshar et al."))

non_kooi_synonymous <- 
  non_kooi_mutations %>% 
  remove_noncoding_snvs()
  
non_kooi_nonsynonymous <- 
  non_kooi_mutations %>% 
  dplyr::anti_join(non_kooi_synonymous)

glue("Of these, {nrow(kooi_mutations)} (Kooi et al., 2016b) were filtered for protein effects with at least two pathogenicity predicting programs whereas {nrow(non_kooi_mutations)} were unfiltered and included {nrow(non_kooi_nonsynonymous)} protein-altering and {nrow(non_kooi_synonymous)} synonymous or non-coding mutations ")
```

```{r}
glue("Past WES/WGS of {sum(n_ngs_samples$n)} treatment-naïve retinoblastomas revealed recurrent secondary mutations or focal deletion in BCOR (in {n_BCOR_prior}) and CREBBP (in {n_CREBBP_prior}) (Supplementary Table S9), and {nrow(private_prior_ngs_snvs)} private somatic mutations ")

```



```{r}
glue("Retention of such variants yielded a total of {sum(stachelek_filtered_tumor_variants)} somatic mutations in 12 tumors (mean {round(mean(stachelek_filtered_tumor_variants), 2)}; median {round(median(stachelek_filtered_tumor_variants), 2)}) (Fig. 4A, circled symbols)")
```

```{r}
glue("")
```


```{r}
glue("of which {english(nrow(filtered_circle_ids$positive))} were subclonal variants detected based on their enrichment and detection in a matched cell line")

```

```{r}
PAN2_vaf <- dplyr::filter(stachelek_filtered_variants, gene == "PAN2") %>%
  dplyr::distinct(sample_type, .keep_all = TRUE) %>%
  dplyr::pull(VAF) %>%
  round(2) %>%
  identity()

SAMD9_vaf <- dplyr::filter(stachelek_filtered_variants, gene == "SAMD9")$VAF %>% 
  round(2)

NAF1_vaf <- dplyr::filter(stachelek_filtered_variants, gene == "NAF1")$VAF %>% 
  round(2)

BCOR28_vaf <- dplyr::filter(stachelek_filtered_variants, gene == "BCOR", alt != "A")$VAF %>% 
  round(2) %>%
  identity()

BCOR33_vaf <- dplyr::filter(stachelek_filtered_variants, gene == "BCOR", alt == "A") %>% 
  dplyr::distinct(sample_type, .keep_all = TRUE) %>% 
  dplyr::pull(VAF) %>%
  round(2) %>%
  identity()


```

```{r, eval = TRUE}
glue("Among the {sum(stachelek_filtered_tumor_variants)} somatic variants detected in CHLA-VC-RB tumors, five affected four genes that were altered in at least one other previously sequenced retinoblastoma sample, including BCOR in two samples and PAN2, NAF1, and SAMD9 in one sample each (Table S1, Fig 5).")
```


```{r, eval = TRUE}
```


```{r, eval = TRUE}
glue("Among the {sum(stachelek_filtered_cell_line_only_variants)} somatic variants detected only in CHLA-VC-RB cell lines, only one (ATM) was in a gene in which variants had been detected in prior retinoblastoma sequencing.")
```


```{r}

loadd(reynolds_scna) 
reynolds_scna <- 
  reynolds_scna %>% 
  # dplyr::filter(study == "Stachelek et al.") %>% 
  # dplyr::filter(str_detect(id, "[0-9]{3}\\.CL")) %>% 
  # dplyr::rename(SYMBOL = symbol) %>%
  identity()
  
reynolds_mutations <- list(snv = reynolds_snv 
                           # scna = reynolds_scna
                           ) %>% 
  purrr::map(dplyr::select, -strand) %>%
  dplyr::bind_rows(.id = "modality") %>%
  identity()

reynolds_variants_per_sample <- 
  reynolds_snv %>% 
  dplyr::group_by(sample) %>% 
  dplyr::count() %>% 
  tibble::deframe()

recurrent_reynolds_mutations <-
  reynolds_mutations %>%
  dplyr::filter(SYMBOL %in% prior_study_snvs$gene) %>%
  # dplyr::filter(is.na(seg.mean)) %>%
  # dplyr::pull(SYMBOL) %>%
  # n_distinct() %>%
    identity()

previously_recurrent_snvs <- 
    noncoding_all_study_snvs %>% 
    dplyr::filter(!str_detect(sample, "-CL")) %>% 
    dplyr::filter(gene %in% recurrent_ngs_genes) %>% 
  dplyr::group_by(gene) %>% 
  dplyr::filter("Stachelek et al." %in% study) %>% 
  dplyr::filter(dplyr::n() > 1) %>% 
  identity()

```

```{r, eval = TRUE}

reynolds_recurrences <- 
  reynolds_mutations %>% 
  dplyr::semi_join(all_study_mutations, by = c("SYMBOL" = "gene"))

recurrent_reynolds_genes <- reynolds_recurrences$SYMBOL

reynolds_internal_recurrences <- 
  reynolds_mutations %>% 
  dplyr::distinct(SYMBOL, sample) %>% 
  dplyr::group_by(SYMBOL) %>%
  dplyr::filter(dplyr::n() > 1) %>%
  dplyr::pull(SYMBOL) %>%
  dplyr::n_distinct() %>%
  identity()

glue("In the 15 unmatched CHLA-RB cell lines we detected novel variants in {n_distinct(reynolds_mutations$Gene)} genes (mean {round(n_distinct(reynolds_mutations$Gene)/15, 2)} per sample) (Table S8), a higher number than for the CHLA-VC-RB cell lines likely reflecting the lack of filtering of private germline polymorphisms. The mutations affected BCOR, and {length(unique(recurrent_reynolds_genes[recurrent_reynolds_genes != \"BCOR\"]))} other genes that were also mutated in other sequenced retinoblastoma samples and are thus candidate recurrent tumor variants (Table S8).  {reynolds_internal_recurrences} genes exhibited novel variants in more than one CHLA-RB sample and one (CDK13) also was mutated in an two CHLA-VC-RB cell lines. ")
```


```{r, eval = TRUE}
glue("The {english(n_distinct(recurrent_ngs_genes))} recurrently altered genes identified by whole exome or whole genome sequencing is significantly more than the n# of recurrences expected by chance among {nrow(all_ngs_snvs)} variants identified in 97 reported retinoblastoma sequences.")


```

```{r}

glue("Gene variants detected in 15 CHLA-RB cell lines included potentially pathogenic mutations in {n_distinct(reynolds_mutations$Gene)} genes (mean {round(mean(reynolds_variants_per_sample), 2)} per sample), a higher number than for the CHLA-VC-RB cell lines likely reflecting the lack of filtering of private germline polymorphisms. The mutations affected BCOR, CREBBP and {n_distinct(recurrent_reynolds_mutations$Gene)} other genes that were mutated in other sequenced retinoblastoma samples and are candidate recurrent tumor variants (Supplementary Table S7)")
```

```{r}

glue("This approach identified {sum(stachelek_filtered_tumor_variants)} potentially pathogenic nucleotide variants (mean {round(mean(stachelek_filtered_tumor_variants),2)}) per tumor, including {nrow(filtered_circle_ids$positive)} variants that were detected by variant callers only in tumor-derived cell lines and subsequently identified in their matched tumors (Fig. 4). The findings are consistent with the low mutation rate (0.0646/Mb) and low numbers of potentially pathogenic somatic variants besides RB1 in prior studies ")

glue("The distribution of VAFs in the current study (mean 3.17) skewed to significantly lower values than prior studies (Fig. 3A), likely reflecting deeper sequencing and rescue of variants initially detected only in matched cell lines.")
```

```{r}
num_vars_cl_only <- sum(stachelek_filtered_cell_line_only_variants)

num_vars_cl <- sum(stachelek_filtered_cell_line_variants)

num_vars_cl_t <- num_vars_cl - sum(stachelek_filtered_cell_line_only_variants)

num_vars_t <- sum(stachelek_filtered_tumor_variants)

cell_line_retention_percentage <- 
  percent(num_vars_cl_only/sum(stachelek_filtered_cell_line_variants))

cl_in_t_percentage <- 
  percent(num_vars_cl_t/num_vars_t)

t_in_cl_percentage <- 
  percent(num_vars_cl_t/num_vars_cl)


glue("This approach also revealed {sum(stachelek_filtered_cell_line_variants)} mutations in the 12 cell lines (mean: {round(mean(stachelek_filtered_cell_line_variants), 2)}; median: {round(median(stachelek_filtered_cell_line_variants))}) of which {num_vars_cl_t} were detected in matched tumors. Thus, {num_vars_cl_t}/{num_vars_t} ({cl_in_t_percentage}) of variants detected in tumors were retained in cell lines and {num_vars_cl_t}/{sum(stachelek_filtered_cell_line_variants)} ({t_in_cl_percentage}) of mutations detected in cell lines had detectable antecedents in their primary tumors.")

```

```{r}

glue("Given the small number of affected genes identified in all analyses ({n_distinct(noncoding_all_study_snvs$gene)}, Table S1), recurrence is strong evidence of a progression-related role. ")
```

```{r}

loadd(noncoding_all_study_snvs)
```

```{r}

consequences <- list(
  nonsynonymous = c("?", "coding_sequence_variant", "frameshift_variant", "inframe_deletion", "inframe_insertion", "missense_variant", "splice_acceptor_variant", "splice_donor_variant", "splice_region_variant", "stop_gained"), 
  UTR = c("three_prime_UTR", "five_prime_UTR", "3_prime_UTR_variant", "5_prime_UTR_variant"), 
  synonymous = c("synonymous_variant", "exon", "stop_retained_variant", "intron_variant", "non_coding_transcript_exon_variant", NA)
) %>% 
  tibble::enframe("consequence_class", "consequence") %>% 
  unnest(cols = c(consequence)) %>% 
  identity()

loadd(all_study_mutations_formatted)


labeled_vars_df <- dplyr::left_join(all_study_mutations_formatted, consequences, by = "consequence") %>% 
  dplyr::filter(modality != "focal_scna") %>% 
  dplyr::filter(!str_detect(sample, "CL")) %>% 
  dplyr::filter(!study == "Afshar et al.") %>% 
  dplyr::group_by(consequence_class)

labeled_vars <-  
  labeled_vars_df %>% 
  dplyr::count() %>%
  tibble::deframe() %>% 
  map(1) %>% 
  identity()
```

```{r}

# "Our analyses included all somatic mutations identified in this and three prior retinoblastoma WES/WGS studies (Kooi et al., 2016b; McEvoy et al., 2014; Zhang et al., 2012) 
glue("and were performed separately for the {labeled_vars$nonsynonymous} variants affecting protein amino acid sequences and for all {nrow(labeled_vars_df)} exomic variants, including {labeled_vars$synonymous} synonymous and {labeled_vars$UTR} UTR sequences of protein coding genes reported in this and two prior studies")
```

```{r}

loadd(displayed_ontologies_table)

displayed_ontologies_genes <- 
  displayed_ontologies_table %>% 
  dplyr::pull(userId) %>% 
  str_split(";") %>% 
  unlist() %>% 
  unique() %>% 
  identity()

displayed_gene_overlap <- 
  labeled_vars_df %>% 
  dplyr::filter(gene %in% displayed_ontologies_genes) %>% 
  dplyr::group_by(consequence_class) %>% 
  dplyr::count() %>% 
  tibble::deframe() %>% 
  purrr::map(1)

displayed_gene_overlap$synonymous <- sum(displayed_gene_overlap$synonymous, displayed_gene_overlap$UTR)

labeled_vars$synonymous <- sum(labeled_vars$synonymous, labeled_vars$UTR)

nonsynonymous_overlap <- percent(displayed_gene_overlap$nonsynonymous/labeled_vars$nonsynonymous)

synonymous_overlap <- percent(displayed_gene_overlap$synonymous/labeled_vars$synonymous)

glue("Remarkably, the significance of four of the top six over-represented terms increased when non-protein-altering variants were included, whereas covalent chromatin modification was less significantly over-represented (Table 2). Overall, {displayed_gene_overlap$nonsynonymous} of {labeled_vars$nonsynonymous} ({nonsynonymous_overlap}) protein-altering variants and {displayed_gene_overlap$synonymous} of {labeled_vars$synonymous} ({synonymous_overlap}) non-protein-altering variants fell in at least one of these six ontology categories.")
```



```{r}

loadd(webgestalt_plot_input)
loadd(noncoding_webgestalt_input)

webgestalt_genes <- 
  webgestalt_plot_input %>% 
  dplyr::pull(userId) %>%
  str_split(";") %>% 
  unlist() %>% 
  unique() %>% 
  identity()

kooi_webgestalt <- 
  noncoding_webgestalt_input %>% 
  dplyr::filter(study == "Kooi et al.") %>% 
  dplyr::filter(gene %in% webgestalt_genes)

loadd(kooi_scnas)
n_kooi_scna_samples <- n_distinct(kooi_scnas$id)

stachelek_webgestalt <- 
  noncoding_webgestalt_input %>% 
  dplyr::filter(study == "Stachelek et al.") %>% 
  dplyr::filter(gene %in% webgestalt_genes)

glue("in a study from Kooi et al. {nrow(kooi_webgestalt)} variants in {n_distinct(kooi_webgestalt$sample)} out of {deframe(n_samples)['Kooi']} samples contributed to enriched ontologies compared to {n_kooi_scna_samples} with recurrent 'RB SCNAs'. While in the present study, {nrow(stachelek_webgestalt)} variants in {n_distinct(stachelek_webgestalt$sample)} out of 12 samples contributed to enriched ontologies compared to 12 with recurrent 'RB SCNAs'")
```

```{r}

loadd(prior_study_snvs_list)
loadd(prior_study_scna_list)

prior_study_snvs_list <- 
  prior_study_snvs_list[names(prior_study_scna_list)] %>% 
  map(mutate, id = sample)

all_altered_samples <- map2(prior_study_snvs_list, prior_study_scna_list, ~union(.x$sample, .y$sample))

myqcs <- 
all_study_snv_qc$vars_per_study %>% 
  split(.$sample_set) %>% 
  map("sample")

mcevoy_wgs_samples <- 
    c(
    "SJRB011_D",
    "SJRB014_D", 
    "SJRB016_D",
    "SJRB020_D", 
    "SJRB024_D",
    "SJRB031_D", 
    "SJRB032_D",
    "SJRB035_D", 
    "SJRB039_D",
    "SJRB051_D"
    )

mcevoy_unaffected_samples <-
  setdiff(mcevoy_wgs_samples, all_altered_samples$mcevoy)

kooi_unaffected_samples <- 
  setdiff(all_altered_samples$kooi, myqcs$Kooi)

zhang_unaffected_samples <- 
  all_altered_samples$zhang %>% 
  str_subset("SJRB*") %>% 
  setdiff(myqcs$Zhang)

all_unaffected_samples <- 
  list(
    mcevoy = mcevoy_unaffected_samples,
    kooi = kooi_unaffected_samples,
    zhang = zhang_unaffected_samples
  ) %>% 
  tibble::enframe("study", "sample") %>% 
  tidyr::unnest() %>% 
  # write_csv("../../doc/dflow_output/unaffected_tumor_samples.csv") %>% 
  identity()
  

```

# tabulate number of focal RB deletions

```{r}

loadd(stachelek_scna_by_gene)

stachelek_rb_scna <- 
  stachelek_scna_by_gene %>% 
  dplyr::filter(symbol == 'RB1', seg.mean < 1.5)

stachelek_rb_scna %>% 
  dplyr::mutate(cohort = ifelse(str_detect(sample_id, "\\.CL"), "reynolds", "vc"), .before = sample_id) %>% 
  dplyr::mutate(sample_type = ifelse(str_detect(sample_id, "CL"), "cell_line", "tumor"), .before = sample_id) %>%
  group_by(cohort, sample_type) %>% 
  dplyr::summarize(n = n_distinct(sample_id))

```